### PR TITLE
WinHttp: always read HTTP/2 streams to the end

### LINKED ISSUE #60291: Subsequent Grpc calls with WinHttpHandler to an IIS hosted service fail
### Description

I'm trying to call a Grpc service hosted in IIS over http2 from NetFramework app. After reading the relevant docs([supported platforms](https://docs.microsoft.com/en-us/aspnet/core/grpc/supported-platforms?view=aspnetcore-5.0#net-grpc-client-requirements)) and issues (https://github.com/dotnet/aspnetcore/issues/9020, https://github.com/dotnet/core/issues/5713) I understand that this should be possible if I'm using Win11/WinServer2022 with the pre release version of the System.Net.Http.WinHttpHandler nuget package (in my case 6.0.0-rc.1.21451.13).

Unfortunately this seems not to be the case, as only the first call to the service succeeds and any subsequent call fails with `Status(StatusCode="Cancelled", Detail="No grpc-status found on response. Using gRPC with WinHttp has Windows and package version requirements. See https://aka.ms/aspnet/grpc/netstandard for details.")`.

Things get even stranger when I tested combining calls made with the NetCore's SocketsHttpHandler. A call with the WinHttpHandler will fail no matter whether the first call to the service was made with WinHttp or Sockets, and it appears that the problem is somehow triggered by the state of the IIS application as restarting the host resets the issue(restarting the client has no effect). Furthermore even after the WinHttp call has failed one can successfully call the service with the Sockets handler from the same process.

Here is a table with the different setups
| Client        | Server        | Result |
| ------------- | ------------- | ------ |
|WinHttp | IIS | :x: Fail |
|WinHttp | Kestrel | :heavy_check_mark: Succeeds |
|Sockets | IIS | :heavy_check_mark: Succeeds |
|Sockets | Kestrel | :heavy_check_mark: Succeeds |

### Reproduction Steps

I've created a minimal project that reproduces the problem - https://github.com/angelyordanov/grpc-iis-winhttp-sample

1. Start the GrpcService with the IISExpress profile
2. Start either one of the console apps

### Expected behavior

There should be no exceptions and all grpc calls should succeed.

### Actual behavior

The second call made with the WinHttpHandler always fails.

### Regression?

No

### Known Workarounds

No

### Configuration

```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.401
 Commit:    4bef5f3dbf

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.401\
```

### Other information

_No response_

### PR DESCRIPTION
**Pull request by antonfirsov**
By it's default behavior, WinHttp stops reading the stream when `Content-Length` is specified, this prevents us to read the remaining trailers. By opting in into `WINHTTP_OPTION_REQUIRE_STREAM_END`, `WinHttpHandler` behavior becomes equivalent of `SocketsHttpHandler`. Note that this means if the actual data is longer than `Content-Length`, we will read all the extra data bytes in both handlers, which we probably shouldn't do, but that's a topic we should investigate separately.

I don't fully understand the impact of applying `WINHTTP_OPTION_REQUIRE_STREAM_END` in HTTP/1.1, so I'm applying it only for HTTP 2.0 requests.

The PR also adds equivalent `Content-Length` tests for SocketsHttpHandler for consistency.

Fixes #60291

/cc @JamesNK

### CODE REVIEW

**Comment by JamesNK** [comment]
Stream end is a http2 and http3 concept. I assume it has no impact on http11.

I would like a unit test of a http2 server returning more content than its content-length defines. We can check what socketshttphandler does, what winhttphandler does today (pre-pr) and what it does after adding this flag.

My experience is in the server side of http, but we validate the content length and error if doesn’t match the actual content.

@tratcher

**Comment by Tratcher** [comment]
The HTTP/2 spec says having a Content-Length mismatch the total data frames is a stream level protocol violation and should result in RST. It's worth confirming WinHttp enforces that.

**Comment by antonfirsov** [comment]
Since this PR is a candidate for our backport process I'd like to stick to the minimal changes necessary to fix #60291. Opened #62906 to track our lack of response `Content-Length` enforcement, current behavior is described in that issue.

**Comment by JamesNK** [comment]
Ok! As long as the impact of the setting is understood then I’m fine with looking at content-length separately.

**Review by JamesNK — APPROVED** [review]

**Comment by antonfirsov** [comment]
@JamesNK I pushed b4b70f0 to set the option right on the stream handle, since the WinHttp team confirmed it doesn't have effect on Http1.1.

If you are fine with the current state, I'm happy to merge this.

**Review by JamesNK — APPROVED** [review]
:shipit:

**Comment by antonfirsov** [comment]
Outerloop failures are unrelated: #63008 #63009

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 1]
```diff
@@ -1145,6 +1146,27 @@ private unsafe void SetTcpKeepalive(SafeWinHttpHandle sessionHandle)
...
+                // According to the WinHttp team, the feature-detection logic in WinHttpTrailersHelper.OsSupportsTrailers
+                // should also indicate the support of WINHTTP_OPTION_REQUIRE_STREAM_END.
+                // WINHTTP_OPTION_REQUIRE_STREAM_END doesn't have effect on HTTP 1.1 requests, therefore it's safe to set it on
+                // the session handle so it is inhereted by all request handles.
+                uint optionData = 1;
+                if (!Interop.WinHttp.WinHttpSetOption(sessionHandle, Interop.WinHttp.WINHTTP_OPTION_REQUIRE_STREAM_END, ref optionData))
+                {
+                    if (NetEventSource.Log.IsEnabled())
+                    {
+                        NetEventSource.Info(this, "Failed to enable WINHTTP_OPTION_REQUIRE_STREAM_END error code: " + Marshal.GetLastWin32Error());
```
Comment (stephentoub) [inline src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:1164]:
```C#
NetEventSource.Info(this, $"Failed to enable WINHTTP_OPTION_REQUIRE_STREAM_END error code: {Marshal.GetLastWin32Error()}");
```

**Pull request merged**